Load subscriptions base and update `is_subscriptions_enabled()` to check for either subscription version

mattallan opened the pull request:
Internal link: paJDYF-2gd-p2

---

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

 - **Load the `subscriptions-base` files when WooCommerce Subscriptions isn't active**

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

In 272a1dbf324bb474040422901c6b55714638a516, I added the `subscriptions-base` files into WooCommerce Payments. These files can be found in `woocommerce-payments/includes/subscriptions/subscriptions-base`.

In this PR, I'm adding a new public function to `woocommerce-payments.php` to init/load the `subscriptions-base` plugin (91ec10e). This function should only load the subscriptions-base when the main WooCommerce Subscriptions extension is not active and we're not in the middle of a request to activate the Subscriptions extension.

Now that WC Pay may load the `subscriptions-base`, we need to update a few functions in `WC_Payment_Gateway_WCPay_Subscriptions_Trait` which was using `WC_Subscriptions::$version` to get the subscriptions version and check if subscriptions is enabled. To reduce duplicate code I added a new `get_subscriptions_version()` function to the subscriptions util trait (7e59bfe) and updated `is_subscriptions_enabled()` function to use this new util function instead (0a101e6)

- **Modify phpcs and lint:css to not run on subscriptions-base**

When trying to push up the `subscriptions-base` there was a lot of phpcs/lint:css issues so I've added a new line and file to ignore these files. Once subscriptions base is in a separate repository we'll have our own coding standing checks.

---

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

**Before testing checkout branch`stripe_billing/load_subscriptions_base`.**

Test case 1: `subscriptions-base` is loading when WC Subscriptions is deactivated

1. Deactivate the WooCommerce Subscriptions plugin
3. Go to the Subscriptions Settings: **WooCommerce > Settings > Subscriptions**
4. You should only see basic subscription settings available (mixed checkout settings and sync [screenshot](https://d.pr/i/u9tv4V))
4. Check the subscriptions table is loading: **WooCommerce > Subscriptions**


Test case 2: `subscriptions-base` is _not_ loaded when WC Subscriptions is active

1. Activate the WooCommerce Subscriptions plugin
2. Check all the Subscriptions settings are loading

Test case 3: basic subscriptions functionality is provided with `subscriptions-base`

1. Make sure WooCommerce Subscriptions plugin is deactivated
2. Add a subscription product to your cart
3. Complete the checkout process
4. Confirm a subscription was created

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


-------------------

- [ ] Added changelog entry (or does not apply)
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

chickenn00dle reviewed (commented):
Thanks for getting this added @mattallan 🙌 

A few notes while I was testing:

- When I deactivated WC Subscriptions, the following admin notice appeared: https://cloudup.com/ch30us6uPzw

I'm assuming this is because we have yet to move base into its own repo and create the package for it, but I wanted to surface this here while I saw it.

- The [subscriptions cart helper](https://github.com/Automattic/woocommerce-payments/blob/develop/tests/unit/helpers/class-wc-helper-subscriptions-cart.php) for unit tests is conflicting with the corresponding class in subscriptions base. It looks like there are a few other subscriptions related helpers we may need to look into as well.
- Subscription base directory is not being ignored for PHP linting checks. See https://github.com/Automattic/woocommerce-payments/pull/2706/checks

mattallan commented:
Thanks for taking a look at this Raz!

> When I deactivated WC Subscriptions, the following admin notice appeared

Hmm, I can't replicate this, but that might be because I'm running 4-0-bleeding of subscriptions.

The `subscriptions-base` in  is set to 4.0.0 [here](https://github.com/Automattic/woocommerce-payments/blob/272a1dbf324bb474040422901c6b55714638a516/includes/subscriptions/subscriptions-base/includes/class-wc-subscriptions-base-plugin.php#L20) so the only way I can get that message to display is if I activate WooCommerce Subscriptions when on `trunk` after the `subscriptions-base` has already been loaded, and then deactivate subscriptions.

This is intended.

>  It looks like there are a few other subscriptions related helpers we may need to look into as well.

Yeah I noticed this too 😭  Now that some subscription classes are being loaded by WC Payments, we may need to remove or rethink how some of those subscription helpers are used in the unit tests.

I'll take a look a this tomorrow.

> Subscription base directory is not being ignored for PHP linting checks

Ahh bugger! I see it's failing the psalm checks so I've just added the subscriptions-base to the list of ignore files and fixed up the other errors: 7695de27

mattallan commented:
@chickenn00dle, I've pushed up a few fixes to the failing checks.

- 8bbbec8 **Fixed phpunit errors**

I spent a while trying to fix the unit tests but in the end, I ended up falling back to not loading the `subscriptions-base` during the tests. I figured this would be okay since we'll have our own unit tests in `subscriptions-base` repo.

The main issues with the unit tests were:
 - WCPAY has created their own versions of many subscriptions public functions, their own`WC_Subscriptions` class and a mocked version of `WC_Subscriptions_Cart` class. The majority of unit tests were failing because the `subscription-base` was loading these classes and functions when they already existed.
   - To work around the way the unit tests are using `WC_Subscriptions_Cart` would require us to mock the static methods which is not easily done.
   - As for working around the public/global wcs functions mocked in the unit tests, the only way would be to wrap all our `wcs-x-functions.php` public functions in subscriptions base with `if function_exists()` checks.
 - Once I worked through all errors related to redeclaring classes/functions, some tests were failing because of x function was expected to run once, but it ran more than once. This error was happening because of all the subscriptions base files/classes/hooks were being loaded and some tests in WCPAY are written to expect a function call once.

- 3cbfc6f **Fixed code coverage dropping below 10%**

I excluded the subscriptions base from the code-coverge 😅 

---

Before merging this, I think @james-allan wants to look at the new `get_subscriptions_version()` function and how it's used in `is_subscriptions_enabled()` so maybe hold off merging :)

chickenn00dle commented:
Thanks for digging into this @mattallan!

>  I ended up falling back to not loading the subscriptions-base during the tests. I figured this would be okay since we'll have our own unit tests in subscriptions-base repo.

This works for me, although I suppose this means we will have to create further subscriptions test classes with Stripe Billing additions.

> To work around the way the unit tests are using WC_Subscriptions_Cart would require us to mock the static methods which is not easily done.

I haven't looked too much into this, but would something like extending Subscriptions base classes in tests, and using these child classes to define static methods, or manipulate behavior work?

> As for working around the public/global wcs functions mocked in the unit tests, the only way would be to wrap all our wcs-x-functions.php public functions in subscriptions base with if function_exists() checks.

Again, I haven't looked too far into this, but I wonder how necessary it is to re-define wcs functions only for WCPay tests. In other words, is it possible to have each individual test suite mock the behavior expected from Subscriptions methods and functions? I'm not sure how much effort would be involved here, but just quickly glancing at the tests directory, I don't see too many subscriptions specific cases.

Overall, I think it's fine to not test Subscriptions base itself in WCPay since it will have its own tests, but considering Stripe Billing relies heavily on this package, it feels strange to have to create test classes in order to write unit tests for it.

chickenn00dle commented:
> Yeah. I had a quick look through our existing stripe billing codebase and the only WCS classes/functions we use are

That's reassuring. I hadn't looked through this yet and was thinking there would be a lot more to mock. Given this isn't the case, I guess the issues I raised aren't really a concern after all.

> I'm pretty nooby when it comes to PHPUnit 

Me too 😞

> I don't think we want this class to be loaded by the subscriptions base because we'll most likely need to mock the values returned by these static function calls (mocking static methods is no longer possible in PHPUnit)

I actually wasn't familiar with the PHPUnit static method limitation until just now, so this all makes a lot more sense.

> I'm sorry! I didn't quite understand this comment :sweat-smile:

I was thinking that rather than define all of these `wcs_` methods globally, it might be possible for each individual test class to create/mock the behavior as needed via something like private methods, mock function, etc. Again, I hadn't looked too much into this, but was just spitballing an idea.

That said, given we don't have to worry too much about not loading base in tests, there isn't really a need to put any effort into this.

mattallan commented:
I'm going to have to get used to running all these checks before pushing the code up 😄

james-allan commented:
LGTM. Thanks @mattallan and @chickenn00dle.
